### ISSUE TITLE
Implement class, subclass, and protocol

### ISSUE THREAD

**Issue by realivanjx** [MelbourneDeveloper/Device.Net#99]
Hi,

Im trying to implement [python-adb project by google](https://github.com/google/python-adb) in my c# project. I noticed that they use usb class, subclass, and protocol to filter out adb devices as described in [this file](https://github.com/google/python-adb/blob/master/adb/adb_commands.py). I cant see them in your `ConnectedDeviceDefinition` class. Are they not yet implemented or placed somewhere else?

Thanks.

**Comment by MelbourneDeveloper** [comment]
@realivanjx great question! 

I have implemented minimal descriptors at this level. You can see where this information is being obtained [here](https://github.com/MelbourneDeveloper/Device.Net/blob/1a5c2073fe2ae2b50fe2bd9ab2fbeae413cf85aa/src/Usb.Net/Windows/WindowsUsbDeviceFactory.cs#L65). It wouldn't be too hard to implement so if you're willing to submit a PR to fix the issue, it would be great. Otherwise, I will look at this for the 3.1 release.

**Comment by realivanjx** [comment]
How abt this?

https://github.com/MelbourneDeveloper/Device.Net/blob/1a5c2073fe2ae2b50fe2bd9ab2fbeae413cf85aa/src/Usb.Net/Windows/USB_DEVICE_DESCRIPTOR.cs

**Comment by realivanjx** [comment]
Will this work?
https://github.com/MelbourneDeveloper/Device.Net/pull/100

**Comment by realivanjx** [comment]
When i tried to build and use the dll from your project i still got these 0s

![image](https://user-images.githubusercontent.com/12814796/62842991-98935780-bce0-11e9-993a-b9d8a87f8954.png)

Did i miss something?

**Comment by MelbourneDeveloper** [comment]
@realivanjx put a breakpoint [here](https://github.com/realivanjx/Device.Net/blob/ba9afbca6e3eea7ee0ac6db4093d33c3b8219d29/src/Usb.Net/Windows/WindowsUsbDeviceFactory.cs#L63) and inspect why the values aren't there:

Have a read of this:
https://docs.microsoft.com/en-us/windows/win32/api/winusb/nf-winusb-winusb_getdescriptor

And this:
https://www.pinvoke.net/default.aspx/Structures.USB_DESCRIPTOR_REQUEST

**Comment by realivanjx** [comment]
this is weird. i created a new console app so i can debug your code and put breakpoint but it is not hitting anything inside `GetDeviceDefinition()`. Do you instead getting usb info somewhere else?

![image](https://user-images.githubusercontent.com/12814796/62855632-f9408580-bd1c-11e9-954d-de24a40c8e78.png)

![image](https://user-images.githubusercontent.com/12814796/62855653-02c9ed80-bd1d-11e9-8615-6aab630a1a32.png)

**Comment by realivanjx** [comment]
Also here is breakpoint on `DeviceManager`:

![image](https://user-images.githubusercontent.com/12814796/62855928-b03d0100-bd1d-11e9-8eb7-1c168108ad71.png)

**Comment by MelbourneDeveloper** [comment]
@realivanjx  you are not awaiting the result. You need to await GetConnectedDeviceDefinitionAsync

**Comment by realivanjx** [comment]
i used `.Result` already

**Comment by realivanjx** [comment]
That field automatically resolves to task's result (replacement of await inside non async function).

![image](https://user-images.githubusercontent.com/12814796/62856226-86d0a500-bd1e-11e9-9915-cea855f59fe3.png)

**Comment by realivanjx** [comment]
In `DeviceManager.GetConnectedDeviceDefinitionsAsync()` your code tries to call `WindowsUsbDeviceFactory.GetConnectedDeviceDefinitionsAsync()` but i dont see that function implementation in `WindowsUsbDeviceFactory` class.

**Comment by realivanjx** [comment]
bingo...

![image](https://user-images.githubusercontent.com/12814796/62907105-df925300-bd9b-11e9-8de7-a469fbe02cb5.png)

**Comment by realivanjx** [comment]
it seems that the function we are trying to debug is never called anywhere

**Comment by MelbourneDeveloper** [comment]
@realivanjx good find. I'm snowed under right now, but should be able to look at this in a couple of days. Do you think you can get your head around the structure enough to fix this issue?

**Comment by realivanjx** [comment]
im just stuck on how to get the usb handle and pass it around

**Comment by realivanjx** [comment]
i have modified `WindowsUsbDeviceFactory` a bit:

```
using Device.Net;
using Device.Net.Exceptions;
using Device.Net.Windows;
using Microsoft.Win32.SafeHandles;
using System;
using System.Runtime.InteropServices;

namespace Usb.Net.Windows
{
    public class WindowsUsbDeviceFactory : WindowsDeviceFactoryBase, IDeviceFactory
    {
        public ushort? ReadBufferSize { get; set; }
        public ushort? WriteBufferSize { get; set; }

        #region Public Override Properties
        public override DeviceType DeviceType => DeviceType.Usb;
        #endregion

        #region Protected Override Methods

        /// <summary>
        /// The parent Guid to enumerate through devices at. This probably shouldn't be changed. It defaults to the WinUSB Guid. 
        /// </summary>
        protected override Guid GetClassGuid() => WindowsDeviceConstants.WinUSBGuid;
        #endregion

        #region Constructor
        public WindowsUsbDeviceFactory(ILogger logger, ITracer tracer) : base(logger, tracer)
        {
        }
        #endregion

        #region Public Methods
        public IDevice GetDevice(ConnectedDeviceDefinition deviceDefinition)
        {
            if (deviceDefinition == null) throw new ArgumentNullException(nameof(deviceDefinition));

            return deviceDefinition.DeviceType != DeviceType ? null : new UsbDevice(new WindowsUsbInterfaceManager(deviceDefinition.DeviceId, Logger, Tracer, ReadBufferSize, WriteBufferSize), Logger, Tracer);
        }
        #endregion

        #region Private Static Methods
        protected override ConnectedDeviceDefinition GetDeviceDefinition(string deviceId)
        {
            // Getting device interface handle.
            SafeFileHandle sfh = APICalls.CreateFile(deviceId, APICalls.GenericWrite | APICalls.GenericRead, APICalls.FileShareRead | APICalls.FileShareWrite, IntPtr.Zero, APICalls.OpenExisting, APICalls.FileAttributeNormal | APICalls.FileFlagOverlapped, IntPtr.Zero);

            if (sfh.IsInvalid)
            {
                //TODO: is error code useful here?
                int errorCode = Marshal.GetLastWin32Error();
                if (errorCode > 0) throw new ApiException($"Device handle no good. Error code: {errorCode}");
            }

            Logger?.Log(Messages.SuccessMessageGotWriteAndReadHandle, nameof(WindowsUsbInterfaceManager), null, LogLevel.Information);

            var isSuccess = WinUsbApiCalls.WinUsb_Initialize(sfh, out var defaultInterfaceHandle);
            WindowsDeviceBase.HandleError(isSuccess, Messages.ErrorMessageCouldntIntializeDevice);

            return GetDeviceDefinition(defaultInterfaceHandle, deviceId);
        }
        #endregion

        #region Public Static Methods
        /// <summary>
        /// Register the factory for enumerating USB devices in Windows.
        /// </summary>
        public static void Register(ILogger logger, ITracer tracer)
        {
            DeviceManager.Current.DeviceFactories.Add(new WindowsUsbDeviceFactory(logger, tracer));
        }

        public static ConnectedDeviceDefinition GetDeviceDefinition(SafeFileHandle defaultInterfaceHandle, string deviceId)
        {
            var deviceDefinition = new ConnectedDeviceDefinition(deviceId) { DeviceType = DeviceType.Usb };

            var bufferLength = (uint)Marshal.SizeOf(typeof(USB_DEVICE_DESCRIPTOR));
            var isSuccess2 = WinUsbApiCalls.WinUsb_GetDescriptor(defaultInterfaceHandle, WinUsbApiCalls.DEFAULT_DESCRIPTOR_TYPE, 0, WinUsbApiCalls.EnglishLanguageID, out var _UsbDeviceDescriptor, bufferLength, out var lengthTransferred);
            WindowsDeviceBase.HandleError(isSuccess2, "Couldn't get device descriptor");

            if (_UsbDeviceDescriptor.iProduct > 0)
            {
                deviceDefinition.ProductName = WinUsbApiCalls.GetDescriptor(defaultInterfaceHandle, _UsbDeviceDescriptor.iProduct, "Couldn't get product name");
            }

            if (_UsbDeviceDescriptor.iSerialNumber > 0)
            {
                deviceDefinition.SerialNumber = WinUsbApiCalls.GetDescriptor(defaultInterfaceHandle, _UsbDeviceDescriptor.iSerialNumber, "Couldn't get serial number");
            }

            if (_UsbDeviceDescriptor.iManufacturer > 0)
            {
                deviceDefinition.Manufacturer = WinUsbApiCalls.GetDescriptor(defaultInterfaceHandle, _UsbDeviceDescriptor.iManufacturer, "Couldn't get manufacturer");
            }

            deviceDefinition.VendorId = _UsbDeviceDescriptor.idVendor;
            deviceDefinition.ProductId = _UsbDeviceDescriptor.idProduct;
            deviceDefinition.WriteBufferSize = _UsbDeviceDescriptor.bMaxPacketSize0;
            deviceDefinition.ReadBufferSize = _UsbDeviceDescriptor.bMaxPacketSize0;
            deviceDefinition.DeviceClass = _UsbDeviceDescriptor.bDeviceClass;
            deviceDefinition.DeviceSubClass = _UsbDeviceDescriptor.bDeviceSubClass;
            deviceDefinition.DeviceProtocol = _UsbDeviceDescriptor.bDeviceProtocol;

            return deviceDefinition;
        }
        #endregion
    }
}
```

It gets better but i still cant get the fields i need:

![image](https://user-images.githubusercontent.com/12814796/62916416-94d80180-bdc2-11e9-99b4-b06fb3ad914b.png)


![image](https://user-images.githubusercontent.com/12814796/62916390-7540d900-bdc2-11e9-877b-21712a3ba27d.png)

**Comment by MelbourneDeveloper** [comment]
@realivanjx can you please add this to the pull request so I can see the diff?

**Comment by realivanjx** [comment]
let me try

**Comment by realivanjx** [comment]
can i include the cli test project too? it is in the same folder as usb.net project

**Comment by MelbourneDeveloper** [comment]
@realivanjx you should put extra projects like that in a separate PR.

**Comment by realivanjx** [comment]
i will remove that once this is done. i will update my pr so u can see the diff.

**Comment by MelbourneDeveloper** [comment]
@realivanjx I can only review small changes one at a time.

**Comment by realivanjx** [comment]
please see this: https://github.com/MelbourneDeveloper/Device.Net/pull/100/commits/75ee22a28e7a15b7f58e1c815ba9417a9fe8da12

**Comment by MelbourneDeveloper** [comment]
@realivanjx , I tried this code: 

```
            deviceDefinition.DeviceClass = _UsbDeviceDescriptor.bDeviceClass;
            deviceDefinition.DeviceSubClass = _UsbDeviceDescriptor.bDeviceSubClass;
            deviceDefinition.DeviceProtocol = _UsbDeviceDescriptor.bDeviceProtocol;
```
But, I just 0s on my device. Do you see anything different on your side?

What's the use case for this? Do you need to filter by these?

**Comment by realivanjx** [comment]
Yes im getting zeroes too. They are for my adb implementation.

**Comment by MelbourneDeveloper** [comment]
@realivanjx are you still interested in fixing this? Otherwise, I will close this.

**Comment by realivanjx** [comment]
I found better implementation of detecting adb devices using class guid so this issue is irrelevant now. Thanks.
